### ISSUE TITLE
AI todo list

### ISSUE THREAD

**Issue by flannery-denny** [bootstrapworld/curriculum#2537]
Before simplifying this list, we had a huge list of noticings about the lessons. Those are recorded in [this document](https://docs.google.com/document/d/1KwbF-FZFJM9zBKpqOwuc73hNdJG7NuBnTVD3ps9yIys/edit?tab=t.nzv6c6y5vpcw), along with notes from meetings with Shriram.  We also have [Recordings of Meetings with Shriram](https://drive.google.com/drive/folders/1my8gOvk56id2n_L2iy8VCtbkanE7y85S)

- [x] add other examples to Data Driven Algorithms and Training lesson 
- Supervised learning: ImageNet, spam tagging, mnest dataset (USPS)
- [x] Figure out where we first introduce the term model. It shouldn't just be defined in the glossary.  - _FD emailed KF with details and hopes she'll resolve._
- [x] We currently only use `@vocab{model}` once in the whole curriculum. That seems wrong.

### SLM2 - Chatbots

**Feedback from SK**
- [x] When we get to the "Predicting the Next Word" part, I'm kinda' confused. I have been clicking all over Soekia, as intended. It would be good to start this by performing a "reset": tell everyone what their settings need to be before they continue. Otherwise you could end up with mayhem as every student is in a different configuration. Agree on the document collection, the N for N-grams, etc.
- [x] It's also confusing because you say (5) unigram then (6) bigram. But if I'm looking at trigrams…why are we talking about uni and bi grams? Hope you can see why a student might be confused about all the Ns floating around.
- [x] «What do you expect Soekia to choose for the third "word". » — should end in a question-mark, not a period.
- [x] • «highest available valid n-gram» — n-gram or N-gram? We seem to be using capital-N in general.
- [ ] • "There is some degree of self-supervision occurring with this type of model." —
1. I really don't like the "some degree of". It's almost all self-supervision.
2. Self-supervision is not a term we have introduced!
3. I don't think we have time to really go into what it means here.
- [x] • "the tic-tac-toe grid is a 3x3 coordinate plane with the origin (0,0) in the bottom left corner" — conceptually, I don't like this phrase. It suggests that (0,0) is inherently at the bottom-left. Rather, it's just a convention we have made up. We could have chosen (1, 1) to be the origin, and we could have put it at the top-right! Etc. This particularly jars with the preceding phrase "knowing that". What we know is that it's a 3x3 grid, and each move must indicate… What we have made up is the convention that (0, 0) is at the bottom-left (and that we are in the first quadrant!).
- [x] • Also, there's a bit missing where we go from "Let's make up a notation" to "How would you annotate". You didn't finish making up the notation! You have to say that the notation is <player><x><y>. Without spelling out the exact syntax, a perfectly reasonable way to write the first move would be "X is at (2, 2)". But it actually really matters for Soekia that we use the much tighter notation we have picked! In fact,  we worked backwards from Soekia to the notation! So I think you have to tell them exact notation ("We will write this as") and give one or two concrete examples first.
- [x] • "Can you translate this "document" written in tic-tac-toe notation" —
1. I would say in "our" ttt notation — it's written to suggest this is some kind of standard notation.
2. Can we put that figure below? Right now it sits very awkwardly off to the side. Also, it could just be plain text…
- [x] • You have "Follow the directions on" and "Complete the first section of". This does not seem to be standard. But I find it quite confusing when I go to a worksheet and then come back to the notes, because I always assume that I should finish the worksheet. It would be much better, I think, if every reference to a worksheet said exactly which section to finish and then come back, but also when to finish it. I see that some other references say this but do so in different ways: having a standard (maybe macro) for worksheet + section would be good, and sometimes the section is "the whole thing" (eg, the "Meet Soekia! (N-grams)" I think is meant to be the whole thing but doesn't say so explicitly.
- [x] • "No, Natural Language Processing - and statistical language modeling, too" — I'm not sure why this is written this exact way, suggesting these are two distinct things.
- [x] • "As long as the language can be broken into "words" that are separated by spaces" — I would take out the "that are separated" part. Yes, Soekia goes by spaces, but it's trivial to change it to use something else as a delimiter. Real language models don't even go by spaces; they use "tokens", which are word-like but not quite words. [https://www.johndcook.com/blog/2025/03/07/tokens-and-words/]

Once SOEKIA is stable:
- [x] Add workbook page were students make predictions based on the valid ngrams with temperature set to zero from the weather collection.
- [x] Add instruction about temperature
- why it's useful 
- a chance for students to notice and wonder about how the suggested words change.
- If the temperature is 60 there’s a 60% chance that we look outside the first set. 100 -> 100% chance. 0 -> 0% chance
- [x] add icons for pause, single step, faster.
- [x] Remove any instruction about Soekia that is tangential
- [x] make sure all instructions with icons and button names match the updated interface

**OLD Feedback that may or may not be relevant to Statistical Language Models and SLM2**
- [ ] "Do you think a text messaging app with a predictive text feature would produce the same results as our class? Why or why not?" there's no need for the @teacher box to be separate from this. The whole thing should be inside the @QandA
- Is the prereq for this lesson really just Training AI 1? I would have thought it needed Training AI 2
- ~There is a google form in this lesson. Could we use mdquiz instead?~ (Moved to 2026 - We would have to be able to give teachers the ability to display the results.)
- [x] remove "statistical phenomena" in the first section because SK doesn't want to define it and it doesn't make sense to use it without a definition
- [ ] Add solutions to "other customizations" section of `tic-tac-toe`
- [x] The chapter  "What is the Role of Language?" seems more like "what makes a language?"

_Sampling from the Language Model_
- Question 1 isn't actually a question, right?
- I like this page overall!

_Soekia: A Closer Look_
- Something about this page feels like it's in the uncanny valley: both over- and under-scaffolded. It's hard to scaffold a mostly-inquiry based activity, so I don't have immediate suggestions here.

### threeish phases of supervised learning - not necessarily sequential
- Demonstration (“supervision”)
- Function abstraction (model construction)
- Use of function (apply the abstraction on new data or study the function)
 
### Introduction to Artificial Intelligence
- [x] add similarity activity - then we can refer back to it in spellchecking, plagiarism, spotify
- how do we measure similarity

### Decision Trees
- [x] Address kathi's feedback. FD

**OLD FEEDBACK that may or may not be relevant**
- Efficiency v. accuracy → talk about "reasoning models" and the challenge of knowing when to reason
-  _Testing and Comparing Decision Trees_
- Won't kids just say the page is wrong because chopsticks aren't on there? I think they need to see one example of this in the lesson before they open this page
- Actually, I'm not clear on what the learning goal of this page is. It feels like there might be two separate goals here: (1) the fact that trees falter on unknown items, and that (2) there are tradeoffs between efficiency and accuracy
- HS-level words "advantageous", "falters", "relatively straightforward"...is "characterizes" an important term (We will use a training dataset that characterizes 14 different shoppers…)

### COMPLETED

- [x] Two big themes we see emerging are quantifying Similarity and that all machine learning relies on data and training.
- [x] rewrite second half of Statistical Language Models so that somebody can attempt to do the activity
- [x] #2602 

**AI Intro SK comments**
- [x] "ChatGPT already appears to be able to…"  — I don't think we need to start with this kind of hagiography… If people know what ChatGPT is they already have an inflated sense of it (and we don't want to appear to be reinforcing it), if they don't, we don't want them to think we think it can do all these things.
- [x] The John McCarthy para — I don't understand why this is a separate para. It should be the second sentence of the previous para. The next para ("According to McCarthy") should also be part of the same para.
- [x] "and then uses statistical models to identify patterns" — they won't know what a "statistical model" is, certainly a middle-school student (or teacher) won't. Why not just "statistics"?
- [x] "which produce better results" — YET AGAIN the same which/that confusion that I have pointed out numerous times. (Yes, I am annoyed.)
- [x] I really don't feel comfortable calling "Google Maps" an "ML technology". What is the ML in (basic) Google Maps?
- [x] Also, why aren't we capitalizing things like "Google Maps" and "Google Translate"?
- [x] I absolutely LOVE the mountain photograph exercise. This is BRILLIANT.  KUDOS. (Btw, I have some pretty nice photos of the Swiss Alps, if you need any…)
- [x] what you really mean is "the amount of space taken up by the mountain/water in the photo". That is, the "units" here are not "meters [height] and square meters [area]" but rather "pixels".
- [x] The transition to AI in pop culture seems rather abrupt. So is the transition to similarity, which even more so shows up out of nowhere. Maybe consider swapping the two? General AI/ML ⭢ AI in Pop Culture is a very smooth transition; maybe there's a natural thing from that lesson to Similarity?
- [x] Can we ask around for newer movies? I swear most of these movies are ancient history to most teens.
- [x] In the Investigate, "AI can perpetuate existing societal biases, leading to unfair outcomes for certain groups" sticks out as being twice as long as everything else, and definitely the one case of excessively pious writing: like we were writing ordinary prose, but felt the need to switch to "woke" mode to make a Profound Point. I would rewrite it to fit into its context.
- [x] QUANTIFYING SIMILARITY "Write a description of Photo 1" — this sounds like a pointless activity. Why not present it as writing alt-text instead? "You have a blind classmate, explain it in as much detail as you can". Same work, but gives it some human context.
- [x] Same issue with mislabeled coordinate planes.

**Data-driven algorithms**
- [x] find and replace FD
- spell check with spell-check
- data driven with data-driven
- [x] Add QandA in replacement of `@teacher{Debrief @printable-exercise{pyret-spell-checker2.adoc} with your students.}`
- ~When this lesson mentions the "what is intelligence?" question, refer back to intro-ai lesson~ - FD disagrees - I think it's best to avoid backward and forward-facing references unless they offer substantive content 
- [x] reformat wb page "Human Spell Checking" So many lines and so much text. FD take a look.
- [x] wait to tell kids that modern ones are way different til after thinking about the first one
- [x] start first spell checker with a none real word instead of "cello"
- [x] change "plausible candidate" to "real-word candidate"
- [x] Reword Questions 5 and 6 to simplify. "Compared to your own strategy when spell-checking, how similar/different…."

_A Pyret Spell Checker: The Algorithm_ Starter File:
-  [x] Starter file should list the possible dictionaries, and an example of using them to spellcheck- RT disagrees
- [x] Replace unicode arrow with `-->`
- [x] remove "edit" from link

_A Pyret Spell Checker: Exploring Different Dataset Sizes_
- [x] This page should have the partially-played game image at the top
- [x] "W1" and "W2" - why are these shown as strings instead of variables?

**Decision Trees**
- [x] remove iris - maybe revisit the examples from earlier in the lesson
- [x] BIG OVERHAUL REQUESTED _Decision Tree: Level 1_ - **FD have a look**
- I bet this could make much more efficient use of space, and possibly include the training dataset!
- "it branches three ways (each with a yes/no) rather than the simple forks…" let's use 2-way branch v. 3-way branch. "Fork" is problematic because forks never have two branches, and because a fork is also a noun from our original decision tree example!
- I'm not sure the overall format of this page works for me. The work done at the top of the page feels useless, since the Y/N answers are given in the middle. 
- [x] _Decision Tree: Level 2_ - Questions 3 and 4 are on the same line

### Training Artificial Intelligence
- [x] Decide what part of Training 2 belong in Training 1 so that it ends with a bigger punch! 
- ~Reframe from scat to percussion? (clap-stomp..)~ MOVED TO 2026
- [x] replace "aggregate" with"collect"
- [x] Case Study: Michelle
- [x] find the @scrub section that needs tightening **FD**

Once this lesson is more stable, see notes document re: specific suggestions for wkbk pages: 
- [x] Designing a Song Recommendation System
- [x] A Primitive Plagiarism Detector
- [x] A Slightly Less Primitive Plagiarism Detector

### Supervised Learning
- [x] should we use "limited course" instead of "closed course" - FD has never heard that term before.
- [x] replace confidence with confidence score and/or measure of confidence
- [x] acknowledge that the actual contract is more complex, including two numbers
- [x] _Supervised Machine Learning: Training ALVINN_
- The first question feels….obvious? And not needing a ton of space for the answer. I could see most of this page being rewritten to ask students to come up with examples of bad training data that misses key properties, or examples of properties that need to be included in a good training set
- I like the "confidence" questions at the bottom!
- [x] Use "training corpus" in this lesson. We use it a lot beforehand, then again in the next lesson. It feels important enough not to ignore
- [x] SK unhappy with this objective: "Recognize that functions can dramatically reduce the complexity of the data to predict the value of the response variable as though it is completely dependent on the explanatory variable." Once reworded, revisit related lesson text.
- [x] add *lossiness*

### Training Artificial Intelligence 2 **FD take a pass**

- [x] Can this vocab be made more friendly? "There is no such protocol with the bag-of-words model. That said, it is crucial to adhere to the same word order for bags of words that you are comparing!"
- [x] Use the terms **dimensionality** and **data normalization** in student-facing text, not just headers
- [x] perhaps add a workbook page for the first section (it is _really_ abstract)
- [x] One strategy PROGRAMMERS? use for this is to represent bags of words as points in space.
Whose strategy is it? (update line 48 with the answer).
- [x] Smooth and/or preview the transition to Pyret in **Computing Closeness with Angle Difference** lesson section
- [x] clearer wording in directions on Human Judgement - assessing the output
- [x] add QandA for debriefing human judgement
- [x] Provide guidance on this question @Q{What happens if we train our plagiarism detector on the internet?}
- [x] _Angle Difference_
- [Specify degrees on this page, and maybe make an optional version that uses radians?
- At least one of these should give the angle, and have students reverse engineer the original strings. Point out that not everyone got the same original strings!
- [x] _The Importance of Human Judgement_
- I wonder if, instead of asking what the numbers mean, we ask them to plug numbers into sentences? 
- Need answer for question 4. Also, why doesn't ask about the number of rays?
- I feel like there's money left on the table here. What would it mean if all the points were truly random? That would be bad, right? Because we expect clumping around Zebra, Soccer, Art, etc. In fact, maybe it really only makes sense to compare within clumps! (which bridges to decision trees)
- [x] _Exploring the Model_ 
- questions 1 and 3 - a table might make this less intimidating
- question 2 - Why? Do we really want to list them all?
- In the lesson plan, I feel like we could get kids to realize the need for cleaning on their own before attempting this page
- [x] Mystery-article undefined!

**Comment by flannery-denny** [comment]
@schanzer @retabak I have finished reading the AI lessons. The comments above contain my preliminary noticings and questions. Please add any thoughts you may have between now and next weeks meeting to the list and we can refine together next week.

**Comment by flannery-denny** [comment]
@schanzer where does this image come from? and do we have permission to use it? google image search didn't return anything identical

![Image](https://github.com/user-attachments/assets/3b3858db-4774-49b0-9217-b0f73fdcd224)

**Comment by schanzer** [comment]
@flannery-denny I made it in chatGPT, and confirmed that we can use it

**Comment by flannery-denny** [comment]
@schanzer wondered if that might have happened. how do we cite it?

**Comment by flannery-denny** [comment]
To share with Shriram on Wednesday:
- SLM and SLM2 - done for now
- Data-driven Algorithms - FD Tackle Human spellchecking reformat and let RT know when done
- Supervised learning

Not ready:
- AI Intro
- Decision Trees (FD tackle the things she committed to in the meantime)
- Training (needs to switch to stomp clap)

**Comment by flannery-denny** [comment]
@schanzer I've put a lot of effort into rethinking and reformatting `decision-trees/part-1` to include the training dataset and fit on a single page. 

For some reason, while the pdf version works,
<img width="869" alt="Image" src="https://github.com/user-attachments/assets/25ffc778-75f0-40d4-aded-75ebaddab440" />

the html version splits this line. 

<img width="921" alt="Image" src="https://github.com/user-attachments/assets/f03e6dbc-5036-4370-9513-e6a288ed1e8c" />

I could ignore the html issue or fix it by making the fitb smaller (I'd rather not) and wanted to flag it as a possible overarching situation.

**Comment by flannery-denny** [comment]
@retabak I'd like to have a call about `decision-trees/part-2` so that we can come to a shared vision before I put more time into it. So that I don't forget and so that you can have a chance to think your way into it before our call... I am confused about:

- why this is only focusing on teens if we also need to complete the tree for people in their thirties.
- why we're using decision stumps here instead of a 2-level decision tree here - I think it would be clearer for me to see the two full stories these trees would tell and using the stump is pushing you to call the decision node the root node, which feels confusing.
